[Outlook] (manifest) event-based activation with the JSON manifest (#3682)
OfficeDev/office-js-docs-pr: docs/outlook/smart-alerts-onmessagesend-walkthrough.md

> - For **SendMode** options available with the `OnMessageSend` and `OnAppointmentSend` events, see [Available SendMode options](/javascript/api/manifest/launchevent#available-sendmode-options).
> - To learn more about manifests for Outlook add-ins, see [Outlook add-in manifests](manifests.md).

# [Teams Manifest (developer preview)](#tab/jsonmanifest)

1. Open the manifest.json file.

1. Add the following object to the "extensions.runtimes" array. Note the following about this markup:

>    - The "minVersion" of the Mailbox requirement set has been set to "1.12" because the table at the top of this article specifies that this is the lowest version of the requirement set that supports the **OnMessageSend** event.

Reviewer: ```suggestion
   - The "minVersion" of the Mailbox requirement set is set to "1.12" because the [supported events table](autolaunch.md#supported-events) specifies that this is the lowest version of the requirement set that supports the `OnMessageSend` event.
```
Proposing use of backticks with the event name to match formatting in events table.

>    - The "id" of the runtime has been set to the descriptive name "autorun_runtime".

Reviewer: ```suggestion
   - The "id" of the runtime is set to the descriptive name "autorun_runtime".
```

>    - The "code" property has a child "page" property that is set to an HTML file and a child "script" property that has been set to a JavaScript file. You will create or edit these files in later steps. Office uses one of these values or the other depending on the platform.

Reviewer: ```suggestion
   - The "code" property has a child "page" property that is set to an HTML file and a child "script" property that is set to a JavaScript file. You'll create or edit these files in later steps. Office uses one of these values depending on the platform.
```

Author: ```suggestion
   - The "code" property has a child "page" property that is set to an HTML file and a child "script" property that is set to a JavaScript file. You'll create or edit these files in later steps. Office uses one of these values or the other depending on the platform.
```

>        - Office on Windows executes the event handler in a JavaScript-only runtime, which must load a JavaScript file directly.

Reviewer: ```suggestion
       - Office on Windows executes the event handler in a JavaScript-only runtime, which loads a JavaScript file directly.
```

>        - Office on Mac and the web execute the handler in a browser runtime, which must load an HTML file. That file, in turn, contains a `<script>` tag that will load the JavaScript file.

Reviewer: ```suggestion
       - Office on Mac and the web execute the handler in a browser runtime, which loads an HTML file. That file, in turn, contains a `<script>` tag that loads the JavaScript file.
```

     For more information, see [Runtimes in Office Add-ins](../testing/runtimes.md).

>    - The "lifetime" property is set to "short" which means that the runtime starts up when the event is triggered and shuts down when the handler completes. (In certain rare cases, the runtime shuts down before the handler completes. See [Runtimes in Office Add-ins](../testing/runtimes.md).)

Reviewer: ```suggestion
   - The "lifetime" property is set to "short", which means that the runtime starts up when the event is triggered and shuts down when the handler completes. (In certain rare cases, the runtime shuts down before the handler completes. See [Runtimes in Office Add-ins](../testing/runtimes.md).)
```

>    - There is an action to run a handler for the **OnMessageSend** event. You will create the handler function in a later step.

Reviewer: ```suggestion
   - There is an action to run a handler for the `OnMessageSend` event. You'll create the handler function in a later step.
```

    ```json
     {
        "requirements": {
            "capabilities": [
                {
                    "name": "Mailbox",
                    "minVersion": "1.12"
                }
            ]
        },
        "id": "autorun_runtime",
        "type": "general",
        "code": {
            "page": "https://localhost:3000/commands.html",
            "script": "https://localhost:3000/launchevent.js"
        },
        "lifetime": "short",
        "actions": [
            {
                "id": "onMessageSendHandler",
                "type": "executeFunction",
                "displayName": "onMessageSendHandler"
            }
        ]
    }
    ```

1. Add the following "autoRunEvents" array as a property of the object in the "extensions" array.

    ```json
    "autoRunEvents": [
    
    ]
    ```

1. Add the following object to the "autoRunEvents" array. Note the following about this code:

>    - The event object assigns a handler function to the **OnMessageSend** event (using the event's Teams manifest name "messageSending" event as described in the table near the top of this article). The function name must match the name used in the "actions" array in an earlier step.

Reviewer: ```suggestion
   - The event object assigns a handler function to the `OnMessageSend` event (using the event's Teams manifest name, "messageSending", as described in the [supported events table](autolaunch.md#supported-events)). The function name provided in "actionId" must match the name used in the "actions" array in an earlier step.
```
Does the actionId name need to match the "id" and/or "displayName" in the action object?

Author: We have to finesse that question for the moment. I believe it is supposed to match "id", but there is a bug that currently makes it necessary to match "displayName". That bug is why we use the same string for "id" and "displayName".

>    - The "sendMode" option has been set to "promptUser". This means that if the message does not meet the conditions that the add-in sets for sending, the user will be prompted to either cancel sending or to send anyway.

Reviewer: ```suggestion
   - The "sendMode" option is set to "promptUser". This means that if the message doesn't meet the conditions that the add-in sets for sending, the user will be prompted to either cancel sending or to send anyway.
```
